src/libdrakvuf/win-processes.c

```diff
@@ -715,6 +715,120 @@ addr_t win_get_wow_peb( drakvuf_t drakvuf, access_context_t* ctx, addr_t eproces
     return ret_peb_addr ;
 }
 
+// see https://github.com/mic101/windows/blob/master/WRK-v1.2/public/internal/base/inc/wow64tls.h#L23
+#define WOW64_TLS_CPURESERVED 1
+
+// magic offset in undocumented structure
+#define WOW64_CONTEXT_PAD 4
+
+status_t win_get_wow_context(drakvuf_t drakvuf, addr_t ethread, addr_t* wow_ctx)
+{
+    access_context_t ctx;
+    ctx.translate_mechanism = VMI_TM_PROCESS_PID;
+    ctx.pid = 4;
+    ctx.addr = thread + drakvuf->offsets[KTHREAD_TEB];
+    addr_t teb_ptr = 0;
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &teb_ptr) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    addr_t eprocess;
+    ctx.addr = thread + drakvuf->offsets[KTHREAD_PROCESS];
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &eprocess) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    pid_t pid;
+
+    if (drakvuf_get_process_pid(drakvuf, eprocess, &pid) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    ctx.translate_mechanism = VMI_TM_PROCESS_DTB;
+
+    if (vmi_pid_to_dtb(drakvuf->vmi, pid, &ctx.dtb) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    addr_t self_teb_ptr;
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &self_teb_ptr) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    addr_t tls_slot;
+    tls_slot = teb_ptr + drakvuf->offsets[TEB_TLS_SLOTS] + (1 * sizeof(uint64_t));
+
+    addr_t tls_slot_2;
+    ctx.addr = tls_slot;
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &tls_slot_2) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    *ret = tls_slot_2 + WOW64_CONTEXT_PAD;
```

Comment:
Because 64 bit TLS slots are not normally used when Windows executes 32 bit aware executables, SysWOW64 uses them to store its internal state. In the second TLS slot there is an undocumented structure `CPURESERVED` in which `CONTEXT` is embedded. I'm not aware of any method that would allow to fetch this offset from Rekall, so magic constant `WOW64_CONTEXT_PAD` is involved here.

Reply:
Checking the [WRK](https://github.com/markjandrews/wrk-v1.2) is always a good idea when in doubt

Reply:
Yes, I even gave a link in one of the comments: https://github.com/mic101/windows/blob/master/WRK-v1.2/public/internal/base/inc/wow64tls.h#L23

The problem is that the constant `WOW64_TLS_CPURESERVED` is just defined in WRK but not actually used anywhere in the repo. It's probably out of scope because a lot of SysWOW is handled in userland.

https://wbenny.github.io/2018/11/04/wow64-internals.html

Here is some good article on the topic:
> RtlWow64GetCpuAreaInfo is an internal ntdll.dll function which is called a lot during emulation. It is mainly used for fetching the machine type and architecture-specific CPU context (the CONTEXT structure) of the emulated process. This information is fetched into an undocumented structure, which we’ll be calling WOW64_CPU_AREA_INFO.

It is somewhere in ntdll but not necessarily in the exported symbols.

Reply:
Hm... https://gist.github.com/wbenny/41b2bf4256f28d61bcc336f10650b3d2#file-2_appendix-h-L318

Looks like it's just a different constant for each architecture, so it would stay `4` as long as it's i386 WOW